types/index.d.ts

```diff
@@ -0,0 +1,88 @@
+// TypeScript Version: 3.5
+
+import {Node, Parent} from 'unist'
+
+declare namespace unistUtilIs {
+  /**
+   * Check that type property matches expectation for a node
+   *
+   * @typeParam T type of node that passes test
+   */
+  type TestType<T extends Node> = T['type']
+
+  /**
+   * Check that some attributes on a node are matched
+   *
+   * @typeParam T type of node that passes test
+   */
+  type TestObject<T extends Node> = Partial<T>
+
+  /**
+   * Check if a node passes a test
+   *
+   * @param node node to check
+   * @param index index of node in parent
+   * @param parent parent of node
+   * @typeParam T type of node that passes test
+   * @returns true if type T is found, false otherwise
+   */
+  type TestFunction<T extends Node> = (
+    node: unknown,
+    index?: number,
+    parent?: Parent
+  ) => node is T
+
+  /**
+   * Union of all the types of tests
+   *
+   * @typeParam T type of node that passes test
+   */
+  type Test<T extends Node> = TestType<T> | TestObject<T> | TestFunction<T>
+}
+
+/**
+ * Unist utility to check if a node passes a test.
+ *
+ * @param test When not given, checks if `node` is a `Node`.
+ * When `string`, works like passing `function (node) {return node.type === test}`.
+ * When `function` checks if function passed the node is true.
+ * When `object`, checks that all keys in test are in node, and that they have (strictly) equal values.
+ * When `array`, checks any one of the subtests pass.
+ * @param node Node to check. `false` is returned
+ * @param index Position of `node` in `parent`
+ * @param parent Parent of `node`
+ * @param context Context object to invoke `test` with
+ * @typeParam T type that node is compared with
+ * @returns Whether test passed and `node` is a `Node` (object with `type` set to non-empty `string`).
+ */
+declare function unistUtilIs<T extends Node>(
+  test: unistUtilIs.Test<T>,
+  node: unknown,
+  index?: number,
+  parent?: Parent,
+  context?: any
+): node is T
+
+/**
+ * Unist utility to check if a node passes a test.
+ *
+ * @param test When not given, checks if `node` is a `Node`.
+ * When `string`, works like passing `function (node) {return node.type === test}`.
+ * When `function` checks if function passed the node is true.
+ * When `object`, checks that all keys in test are in node, and that they have (strictly) equal values.
+ * When `array`, checks any one of the subtests pass.
+ * @param node Node to check. `false` is returned
+ * @param index Position of `node` in `parent`
+ * @param parent Parent of `node`
+ * @param context Context object to invoke `test` with
+ * @returns Whether test passed and `node` is a `Node` (object with `type` set to non-empty `string`).
+ */
+declare function unistUtilIs(
```

Comment:
Sadly it should be impossible to populate types from Test array....
I think the function should still accept generic.

```ts
declare function unistUtilIs<T extends Node>(
  test: Array<unistUtilIs.Test<any>>,
  node: unknown,
  index?: number,
  parent?: Parent,
  context?: any
): node is T
```

Then, adopters still can use type guards.

```ts
let randomNode: unknown
if (is<Heading | Element>([isHeading, isElement], randomNode)) {
  randomNode // Now it is recognized as a Heading | Element.
}
```

How do you think?

Reply:
I'm open to this idea.
The reason I didn't initially is this edge case

```ts
let randomNode: unknown
if (is<RandomType | NotAnElement>([isHeading, isElement], randomNode)) {
  randomNode // Node is a `Heading | Element`, but is typed as `RandomType | NotAnElement`.
}
```

The thinking being that it would be better for adopters to have to narrow the `Node` type again, and have full type safety, than to allow an incorrect type.

Again, that may or may not be a good approach, just my initial thought.

Reply:
> The thinking being that it would be better for adopters to have to narrow the Node type again, and have full type safety, than to allow an incorrect type.

Narrowing the type is definitely required. But I think adopters can do more efficiently with the generics.

```ts
if (is<RandomType | NotAnElement>([isHeading, isElement], randomNode)) {
  // Now we don't need to check again that the randomNode is a valid Node object.
  // So we can distinguish the node by its type.
  switch (randomNode.type) {
    case 'heading':
      ... // randomNode is typed as Heading
    case 'element':
      ... // randomNode is typed as Element
  }
}
```

Reply:
merged the overloaded array of tests version into the generic.
and added some tests showing how types should be used in that case.